### PR TITLE
Boot: Add 'revert' support to direct-xip mode

### PR DESCRIPTION
**Pull request by davidvincze**
The 'revert' mechanism in direct-xip is similar to the one in swap mode.
It also requires the trailer magic to be added to the signed image. When a
reboot happens without the image being confirmed at runtime (without the
image_ok flag being set), the bootloader considers the image faulty and
will attempt to boot the previous image instead. The images can also be
made permanent (marked as confirmed in advance) just like in swap mode.

### CODE REVIEW

File: boot/bootutil/src/loader.c [hunk 1]
```diff
@@ -2037,6 +2041,57 @@ context_boot_go(struct boot_loader_state *state, struct boot_rsp *rsp)
                 }
             }
 
+#ifdef MCUBOOT_DIRECT_XIP_REVERT
```
[the remark this answers is not in the document]
Comment (hakonfam) [reply boot/bootutil/src/loader.c:2044]: I agree with this, the outermost for-loop now almost stretches beyond my monitor :)

File: boot/zephyr/include/mcuboot_config/mcuboot_config.h [hunk 2]
```diff
@@ -70,6 +70,10 @@
 #define MCUBOOT_DIRECT_XIP
 #endif
 
+#ifdef BOOT_DIRECT_XIP_REVERT
```
Comment (hakonfam) [inline boot/zephyr/include/mcuboot_config/mcuboot_config.h:73]:
```suggestion
#ifdef CONFIG_BOOT_DIRECT_XIP_REVERT
```

**Review by hakonfam — CHANGES_REQUESTED** [review]
Also I would update the commit message to state that the image is deleted if its reverted.

**Comment by utzig** [comment]
Please rebase

File: boot/zephyr/Kconfig [hunk 3]
```diff
@@ -204,9 +204,18 @@ config BOOT_DIRECT_XIP
...
 
-if !SINGLE_IMAGE_DFU
+config BOOT_DIRECT_XIP_REVERT
+	bool "Enable the revert mechanism in direct-xip mode"
+	default n
+	help
+	  If y, enables the revert mechanism in direct-xip similar to the one in
+	  swap mode. It requires the trailer magic to be added to the signed image.
+	  When a reboot happens without the image being confirmed at runtime, the
+	  bootloader considers the image faulty and will attempt to boot the
```
[the remark this answers is not in the document]
Comment (davidvincze) [reply boot/zephyr/Kconfig]: Done.

File: boot/bootutil/src/loader.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (davidvincze) [reply boot/bootutil/src/loader.c:2292]: Done.

File: boot/zephyr/include/mcuboot_config/mcuboot_config.h (hunk 2 above)
Comment (davidvincze) [reply boot/zephyr/include/mcuboot_config/mcuboot_config.h]: Fixed.

**Review by hakonfam — APPROVED** [review]

File: boot/bootutil/src/loader.c [hunk 4]
```diff
@@ -2217,6 +2288,17 @@ context_boot_go(struct boot_loader_state *state, struct boot_rsp *rsp)
                     selected_image_header = hdr;
                 }
             }
+
+#ifdef MCUBOOT_DIRECT_XIP_REVERT
+            rc = boot_select_or_erase(slot_state, selected_slot);
```
Comment (utzig) [inline boot/bootutil/src/loader.c:2293]: ...erase(`&slot_state`, ...)

File: boot/bootutil/src/loader.c [hunk 5]
```diff
@@ -2007,6 +2007,75 @@ boot_get_slot_usage(struct boot_loader_state *state, uint8_t slot_usage[],
...
+ */
+static int
+boot_select_or_erase(struct boot_swap_state *state, uint32_t slot)
+{
+    const struct flash_area *fap;
+    int fa_id;
+    int rc;
+
+    fa_id = flash_area_id_from_image_slot(slot);
+    rc = flash_area_open(fa_id, fap);
```
Comment (utzig) [inline boot/bootutil/src/loader.c:2030]: `rc = flash_area_open(fa_id, &fap);`

File: boot/bootutil/src/loader.c [hunk 6]
```diff
@@ -2007,6 +2007,75 @@ boot_get_slot_usage(struct boot_loader_state *state, uint8_t slot_usage[],
...
+    assert(rc == 0);
+
+    memset(state, 0, sizeof(struct boot_swap_state));
+    rc = boot_read_swap_state(fap, &state);
```
Comment (utzig) [inline boot/bootutil/src/loader.c:2034]: `rc = boot_read_swap_state(fap, state);`

**Review by utzig — CHANGES_REQUESTED** [review]
Would not it be better to make this just part of the standard `DIRECT_XIP` instead of adding a new option?

File: boot/bootutil/src/loader.c (hunk 4 above)
Comment (davidvincze) [reply boot/bootutil/src/loader.c]: Fixed. I have no idea what I was doing..

File: boot/bootutil/src/loader.c (hunk 6 above)
Comment (davidvincze) [reply boot/bootutil/src/loader.c]: Fixed.

File: boot/bootutil/src/loader.c (hunk 5 above)
Comment (davidvincze) [reply boot/bootutil/src/loader.c]: Fixed.

**Comment by davidvincze** [comment]
> Would not it be better to make this just part of the standard `DIRECT_XIP` instead of adding a new option?

Without enabling this additional mechanism, it is possible to boot an image without having to write/modify the flash.
Therefore for platform with relatively large write-sizes, the direct-xip mode can be a convenient alternative.

**Comment by nvlsianpu** [comment]
``` consloe
** Booting Zephyr OS build zephyr-v2.4.0-1111-g8941f9a00c0b  ***
[00:00:00.398,529] <inf> mcuboot: Starting bootloader
[00:00:00.404,510] <inf> mcuboot: Primary slot: Image not found
[00:00:00.411,163] <inf> mcuboot: Secondary slot: version=1.1.0+0
[00:00:00.501,007] <inf> mcuboot: Booting image from the secondary slot
[00:00:00.508,361] <inf> mcuboot: Bootloader chainload address offset: 0x73000
[00:00:00.516,357] <inf> mcuboot: Jumping to the first image slot
*** Booting Zephyr OS build zephyr-v2.4.0-1111-g8941f9a00c0b  ***
*** Booting Zephyr OS build zephyr-v2.4.0-1111-g8941f9a00c0b  ***
[00:00:00.392,059] <inf> mcuboot: Starting bootloader
[00:00:00.398,040] <inf> mcuboot: Primary   slot: version=1.2.0+0
[00:00:00.404,876] <inf> mcuboot: Secondary slot: version=1.1.0+0
[00:00:00.493,438] <inf> mcuboot: Booting image from the primary slot
[00:00:00.500,610] <inf> mcuboot: Bootloader chainload address offset: 0xc000
[00:00:00.508,514] <inf> mcuboot: Jumping to the first image slot
*** Booting Zephyr OS build zephyr-v2.4.0-1111-g8941f9a00c0b  ***
Hello World! nrf52840dk_nrf52840
*** Booting Zephyr OS build zephyr-v2.4.0-1111-g8941f9a00c0b  ***
[00:00:00.256,347] <inf> mcuboot: Starting bootloader
[00:00:00.262,329] <inf> mcuboot: Primary   slot: version=1.2.0+0
[00:00:00.269,134] <inf> mcuboot: Secondary slot: version=1.1.0+0
[00:00:17.703,399] <err> mcuboot: Unable to find bootable image
*** Booting Zephyr OS build zephyr-v2.4.0-1111-g8941f9a00c0b  ***
[00:00:00.256,286] <inf> mcuboot: Starting bootloader
[00:00:00.262,268] <inf> mcuboot: Primary slot: Image not found
[00:00:00.268,890] <inf> mcuboot: Secondary slot: Image not found
[00:00:00.275,695] <err> mcuboot: Unable to find bootable image
```

**Comment by utzig** [comment]
@nvlsianpu You might be missing `--pad` when creating images; it seems to check for `MAGIC` and delete if it's not there. A better strategy would be to validate and only erase when images are invalid.

**Review by nvlsianpu — APPROVED** [review]
Approved since it doesn't corrupted what is already working.
Any issues might been fixed after.

**Review by nvlsianpu — CHANGES_REQUESTED** [review]
Need to clarify why it doesn't work for me.

**Comment by davidvincze** [comment]
> @nvlsianpu You might be missing `--pad` when creating images; it seems to check for `MAGIC` and delete if it's not there. A better strategy would be to validate and only erase when images are invalid.

In such a case I'm afraid  you could be left with a valid, but non-working firmware.

**Comment by davidvincze** [comment]
@nvlsianpu it looks like after the third boot it found the image in the primary slot to be invalid.
[00:00:08.995,880] <err> mcuboot: Unable to find bootable image.

Or did the revert code delete the slot? Can you confirm that this image can be loaded? (without performing any upgrades)

**Comment by davidvincze** [comment]
@nvlsianpu  I did perform a similar test and I was working as expected.
- image with higher version was booted from secondary slot,
- after reset it was deleted as it was not confirmed, booted image in primary slot,
- image image primary slot proved to be stable, it was made permanent (marked as confirmed by the imgtool) and 
  I could reset the board and boot this image (several times) and it was not erased by the bootloader

**Comment by nvlsianpu** [comment]
Issue was that I was using imgtool from pip of version 1.6.0 which haven't this patch #806.

**Review by nvlsianpu — APPROVED** [review]

**Review by utzig — APPROVED** [review]

**Pull request merged**